File: fusefs/src/dek-store-v1.js
const crypto = require('node:crypto');
const { sealAes256Gcm, openAes256Gcm, V1: CRYPTO_V1, randomKey32 } = require('./crypto-v1');

// v1 DEK/KEK hierarchy.
//
// - KEK: 32-byte secret stored in macOS Keychain (wrapper responsibility).
// - DEK: 32-byte secret used to encrypt file contents.
// - DEK is stored on disk only in wrapped form (encrypted under KEK).
//
// This module defines a small, versioned "wrapped DEK" format.

const V1 = {
  WRAPPED_DEK_MAGIC: Buffer.from('OCDEK1', 'utf8'),
  WRAPPED_DEK_VERSION: 0x01,
  NONCE_LEN: CRYPTO_V1.NONCE_LEN,
  TAG_LEN: CRYPTO_V1.TAG_LEN,
  DEK_LEN: 32,
};

function newDek() {
  return randomKey32();
}

function encodeWrappedDekV1({ kek, dek }) {
  if (!Buffer.isBuffer(kek) || kek.length !== 32) throw new Error('kek must be 32 bytes');
  if (!Buffer.isBuffer(dek) || dek.length !== V1.DEK_LEN) throw new Error('dek must be 32 bytes');

  const nonce = crypto.randomBytes(V1.NONCE_LEN);
  const header = Buffer.concat([V1.WRAPPED_DEK_MAGIC, Buffer.from([V1.WRAPPED_DEK_VERSION, V1.NONCE_LEN]), nonce]);

  const { ciphertext, tag } = sealAes256Gcm({ key: kek, nonce, aad: header, plaintext: dek });
  return Buffer.concat([header, ciphertext, tag]);
}

function decodeWrappedDekV1({ kek, blob }) {
  if (!Buffer.isBuffer(kek) || kek.length !== 32) throw new Error('kek must be 32 bytes');
  if (!Buffer.isBuffer(blob)) throw new Error('blob must be a Buffer');

  const minLen = V1.WRAPPED_DEK_MAGIC.length + 2 + V1.NONCE_LEN + V1.DEK_LEN + V1.TAG_LEN;
  if (blob.length < minLen) throw new Error('wrapped DEK blob too small');

  const magic = blob.subarray(0, V1.WRAPPED_DEK_MAGIC.length);
  if (!magic.equals(V1.WRAPPED_DEK_MAGIC)) throw new Error('bad wrapped DEK magic');

  const version = blob[V1.WRAPPED_DEK_MAGIC.length + 0];
  const nonceLen = blob[V1.WRAPPED_DEK_MAGIC.length + 1];
  if (version !== V1.WRAPPED_DEK_VERSION) throw new Error(`unsupported wrapped DEK version: ${version}`);
  if (nonceLen !== V1.NONCE_LEN) throw new Error(`unsupported wrapped DEK nonceLen: ${nonceLen}`);

  const headerLen = V1.WRAPPED_DEK_MAGIC.length + 2 + V1.NONCE_LEN;
  const header = blob.subarray(0, headerLen);
  const nonce = blob.subarray(V1.WRAPPED_DEK_MAGIC.length + 2, headerLen);

  const tag = blob.subarray(blob.length - V1.TAG_LEN);
  const ciphertext = blob.subarray(headerLen, blob.length - V1.TAG_LEN);

  const dek = openAes256Gcm({ key: kek, nonce, aad: header, ciphertext, tag });
  if (dek.length !== V1.DEK_LEN) throw new Error('unexpected DEK length');
  return dek;
}

module.exports = {
  V1,
  newDek,
  encodeWrappedDekV1,
  decodeWrappedDekV1,
};

File: fusefs-swift/Sources/OcProtectFsFuse/DekStoreV1.swift
import Foundation
import CryptoKit

// Phase 3 (Issue #109): wrapped DEK store parity with fusefs/src/dek-store.js.
//
// Wrapped DEK format (v1):
//   magic "OCDEK1" (6 bytes)
//   version (1 byte) = 0x01
//   nonceLen (1 byte) = 12
//   nonce (12 bytes)
//   ciphertext (32 bytes)
//   tag (16 bytes)
//
// AEAD AAD = header (magic+version+nonceLen+nonce)

enum DekStoreV1 {
  enum Const {
    static let magic = Data("OCDEK1".utf8)
    static let version: UInt8 = 0x01
    static let nonceLen = 12
    static let dekLen = 32
    static let tagLen = 16
  }

  struct DekError: Error, CustomStringConvertible {
    let description: String
  }

  static func newDek() -> Data {
    var b = [UInt8](repeating: 0, count: Const.dekLen)
    _ = SecRandomCopyBytes(kSecRandomDefault, b.count, &b)
    return Data(b)
  }

  static func encodeWrappedDek(kek: Data, dek: Data, nonce: Data? = nil) throws -> Data {
    guard kek.count == 32 else { throw DekError(description: "kek must be 32 bytes") }
    guard dek.count == Const.dekLen else { throw DekError(description: "dek must be 32 bytes") }

    let useNonce: Data
    if let nonce {
      guard nonce.count == Const.nonceLen else { throw DekError(description: "nonce must be 12 bytes") }
      useNonce = nonce
    } else {
      var b = [UInt8](repeating: 0, count: Const.nonceLen)
      _ = SecRandomCopyBytes(kSecRandomDefault, b.count, &b)
      useNonce = Data(b)
    }

    var header = Data()
    header.append(Const.magic)
    header.append(contentsOf: [Const.version, UInt8(Const.nonceLen)])
    header.append(useNonce)

    let key = SymmetricKey(data: kek)
    let nonce = try AES.GCM.Nonce(data: useNonce)
    let sealed = try AES.GCM.seal(dek, using: key, nonce: nonce, authenticating: header)

    guard sealed.ciphertext.count == Const.dekLen else { throw DekError(description: "unexpected ciphertext length") }
    guard sealed.tag.count == Const.tagLen else { throw DekError(description: "unexpected tag length") }

    var out = Data()
    out.append(header)
    out.append(sealed.ciphertext)
    out.append(sealed.tag)
    return out
  }

  static func decodeWrappedDek(kek: Data, blob: Data) throws -> Data {
    guard kek.count == 32 else { throw DekError(description: "kek must be 32 bytes") }

    let minLen = Const.magic.count + 2 + Const.nonceLen + Const.dekLen + Const.tagLen
    guard blob.count >= minLen else { throw DekError(description: "wrapped DEK blob too small") }

    let magic = blob.subdata(in: 0..<Const.magic.count)
    guard magic == Const.magic else { throw DekError(description: "bad wrapped DEK magic") }

    let version = blob[Const.magic.count + 0]
    let nonceLen = Int(blob[Const.magic.count + 1])
    guard version == Const.version else { throw DekError(description: "unsupported wrapped DEK version: \(version)") }
    guard nonceLen == Const.nonceLen else { throw DekError(description: "unsupported wrapped DEK nonceLen: \(nonceLen)") }

    let headerLen = Const.magic.count + 2 + Const.nonceLen
    let header = blob.subdata(in: 0..<headerLen)
    let nonceData = blob.subdata(in: (Const.magic.count + 2)..<headerLen)

    let tagStart = blob.count - Const.tagLen
    let ciphertext = blob.subdata(in: headerLen..<tagStart)
    let tag = blob.subdata(in: tagStart..<blob.count)

    let key = SymmetricKey(data: kek)
    let nonce = try AES.GCM.Nonce(data: nonceData)
    let sealed = try AES.GCM.SealedBox(nonce: nonce, ciphertext: ciphertext, tag: tag)

    let dek = try AES.GCM.open(sealed, using: key, authenticating: header)
    guard dek.count == Const.dekLen else { throw DekError(description: "unexpected DEK length") }
    return dek
  }
}
